notebook/services/kernels/kernelmanager.py

```diff
@@ -156,38 +304,49 @@ def start_kernel(self, kernel_id=None, path=None, **kwargs):
             an existing kernel is returned, but it may be checked in the future.
         """
         if kernel_id is None:
-            if path is not None:
-                kwargs['cwd'] = self.cwd_for_path(path)
-            kernel_id = yield gen.maybe_future(
-                super(MappingKernelManager, self).start_kernel(**kwargs)
-            )
-            self._kernel_connections[kernel_id] = 0
-            self.start_watching_activity(kernel_id)
-            self.log.info("Kernel started: %s" % kernel_id)
-            self.log.debug("Kernel args: %r" % kwargs)
-            # register callback for failed auto-restart
-            self.add_restart_callback(kernel_id,
-                lambda : self._handle_kernel_died(kernel_id),
-                'dead',
-            )
-
-            # Increase the metric of number of kernels running
-            # for the relevant kernel type by 1
-            KERNEL_CURRENTLY_RUNNING_TOTAL.labels(
-                type=self._kernels[kernel_id].kernel_name
-            ).inc()
-
+            kernel_id = self.start_launching_kernel(path, kernel_name, **kwargs)
+            yield self.get_kernel(kernel_id).client_ready()
         else:
             self._check_kernel_id(kernel_id)
             self.log.info("Using existing kernel: %s" % kernel_id)
 
-        # Initialize culling if not already
-        if not self._initialized_culler:
-            self.initialize_culler()
-
         # py2-compat
         raise gen.Return(kernel_id)
 
+    def start_launching_kernel(self, path=None, kernel_name=None, **kwargs):
+        """Launch a new kernel, return its kernel ID
+
+        This is a synchronous method which starts the process of launching a
+        kernel. Retrieve the KernelInterface object and call ``.client_ready()``
+        to get a future for the rest of the startup & connection.
+        """
+        if path is not None:
+            kwargs['cwd'] = self.cwd_for_path(path)
+        kernel_id = str(uuid.uuid4())
+        if kernel_name is None:
+            kernel_name = 'pyimport/kernel'
+        elif '/' not in kernel_name:
+            kernel_name = 'spec/' + kernel_name
+
+        kernel = KernelInterface(kernel_name, self.kernel_finder)
```

Comment:
We should add kwargs to the `KernelInterface()` constructor.  The gateway projects (Kernel and Enterprise) allow for users to specify arguments that influence kernel creation.  For example, users can specify a pre-existing namespace (via `KERNEL_NAMESPACE`) that instructs Enterprise Gateway to launch kubernetes-based kernels into the specified namespace rather than having EG create one for them.  Since `KernelInterface()` is the "path" to the kernel provider, we should _flow_ `kwargs`.

Reply:
Thanks. I may think more about the MappingKernelManager and KernelInterface APIs, and I'll try to keep this in mind if I do. I'm not entirely happy with them at present.